Subject: [PATCH] gitweb: unify shortlog/log links

Instead of 'shortlog | log', we use only 'log' now, pointing to the
shortlog action, since that is what most people are usually interested
in first.  In the action-specific navbar, you can then choose between
short and full log view.

The aim of this patch is to reduce the overwhelming number of links
by including only orthogonal links to most popular views.

Signed-off-by: Petr Baudis <petr.baudis@novartis.com>
---
 gitweb/gitweb.perl | 48 ++++++++++++++++++++++++++++++++++++------------
 1 file changed, 36 insertions(+), 12 deletions(-)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index 6c8f971a..c5a5fe61 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -4347,7 +4347,7 @@ sub git_print_page_nav {
 	my ($current, $suppress, $head, $treehead, $treebase, $extra) = @_;
 	$extra = '' if !defined $extra; # pager or formats
 
-	my @navs = qw(summary shortlog log commit commitdiff tree);
+	my @navs = qw(summary log commit commitdiff tree);
 	if ($suppress) {
 		@navs = grep { $_ ne $suppress } @navs;
 	}
@@ -4358,12 +4358,16 @@ sub git_print_page_nav {
 			$arg{$_}{'hash'} = $head;
 		}
 		if ($current =~ m/^(tree | log | shortlog | commit | commitdiff | search)$/x) {
-			for (qw(shortlog log)) {
-				$arg{$_}{'hash'} = $head;
-			}
+			$arg{'log'}{'hash'} = $head;
 		}
 	}
 
+	$arg{'log'}{'action'} = 'shortlog';
+	if ($current eq 'log') {
+		$current = 'shortlog';
+	} elsif ($current eq 'shortlog') {
+		$current = 'log';
+	}
 	$arg{'tree'}{'hash'} = $treehead if defined $treehead;
 	$arg{'tree'}{'hash_base'} = $treebase if defined $treebase;
 
@@ -4432,6 +4436,26 @@ sub format_paging_nav {
 	return $paging_nav;
 }
 
+sub format_log_nav {
+	my ($action, $page, $has_next_link) = @_;
+	my $paging_nav;
+
+	if ($action eq 'shortlog') {
+		$paging_nav .= 'shortlog';
+	} else {
+		$paging_nav .= $cgi->a({-href => href(action=>'shortlog', -replay=>1)}, 'shortlog');
+	}
+	$paging_nav .= ' | ';
+	if ($action eq 'log') {
+		$paging_nav .= 'fulllog';
+	} else {
+		$paging_nav .= $cgi->a({-href => href(action=>'log', -replay=>1)}, 'fulllog');
+	}
+
+	$paging_nav .= " | " . format_paging_nav($action, $page, $has_next_link);
+	return $paging_nav;
+}
+
 ## ......................................................................
 ## functions printing or outputting HTML: div
 
@@ -5766,8 +5790,7 @@ sub git_project_list_rows {
 		}
 		print"<td class=\"link\">" .
 		      $cgi->a({-href => href(project=>$pr->{'path'}, action=>"summary")}, "summary")   . " | " .
-		      $cgi->a({-href => href(project=>$pr->{'path'}, action=>"shortlog")}, "shortlog") . " | " .
-		      $cgi->a({-href => href(project=>$pr->{'path'}, action=>"log")}, "log") . " | " .
+		      $cgi->a({-href => href(project=>$pr->{'path'}, action=>"shortlog")}, "log") . " | " .
 		      $cgi->a({-href => href(project=>$pr->{'path'}, action=>"tree")}, "tree") .
 		      ($pr->{'forks'} ? " | " . $cgi->a({-href => href(project=>$pr->{'path'}, action=>"forks")}, "forks") : '') .
 		      "</td>\n" .
@@ -6056,8 +6079,7 @@ sub git_tags_body {
 		      "<td class=\"link\">" . " | " .
 		      $cgi->a({-href => href(action=>$tag{'reftype'}, hash=>$tag{'refid'})}, $tag{'reftype'});
 		if ($tag{'reftype'} eq "commit") {
-			print " | " . $cgi->a({-href => href(action=>"shortlog", hash=>$tag{'fullname'})}, "shortlog") .
-			      " | " . $cgi->a({-href => href(action=>"log", hash=>$tag{'fullname'})}, "log");
+			print " | " . $cgi->a({-href => href(action=>"shortlog", hash=>$tag{'fullname'})}, "log");
 		} elsif ($tag{'reftype'} eq "blob") {
 			print " | " . $cgi->a({-href => href(action=>"blob_plain", hash=>$tag{'refid'})}, "raw");
 		}
@@ -6096,8 +6118,7 @@ sub git_heads_body {
 		               -class => "list name"},esc_html($ref{'name'})) .
 		      "</td>\n" .
 		      "<td class=\"link\">" .
-		      $cgi->a({-href => href(action=>"shortlog", hash=>$ref{'fullname'})}, "shortlog") . " | " .
-		      $cgi->a({-href => href(action=>"log", hash=>$ref{'fullname'})}, "log") . " | " .
+		      $cgi->a({-href => href(action=>"shortlog", hash=>$ref{'fullname'})}, "log") . " | " .
 		      $cgi->a({-href => href(action=>"tree", hash=>$ref{'fullname'}, hash_base=>$ref{'fullname'})}, "tree") .
 		      "</td>\n" .
 		      "</tr>";
@@ -7459,7 +7480,7 @@ sub git_log_generic {
 	}
 
 
-	my $paging_nav = format_paging_nav($fmt_name, $page, $#commitlist >= 100);
+	my $paging_nav = format_log_nav($fmt_name, $page, $#commitlist >= 100);
 	my $next_link = '';
 	if ($#commitlist >= 100) {
 		$next_link =
@@ -7475,7 +7496,10 @@ sub git_log_generic {
 		}
 	}
 
-	git_header_html();
+	{
+		local $action = 'fulllog';
+		git_header_html();
+	}
 	git_print_page_nav($fmt_name,'', $hash,$hash,$hash, $paging_nav);
 	if (defined $file_name) {
 		git_print_header_div('commit', esc_html($co{'title'}), $base);
---
